thirdparty/colmap/scripts/python/read_write_model.py

```diff
@@ -0,0 +1,507 @@
+                id=camera_id, model=model_name, width=width, height=height, params=np.array(params)
+            )
+        assert len(cameras) == num_cameras
+    return cameras
+
+
+def write_cameras_text(cameras, path):
+    """
+    see: src/base/reconstruction.cc
+        void Reconstruction::WriteCamerasText(const std::string& path)
+        void Reconstruction::ReadCamerasText(const std::string& path)
+    """
+    HEADER = (
+        "# Camera list with one line of data per camera:\n"
+        + "#   CAMERA_ID, MODEL, WIDTH, HEIGHT, PARAMS[]\n"
+        + "# Number of cameras: {}\n".format(len(cameras))
+    )
+    with open(path, "w") as fid:
+        fid.write(HEADER)
+        for _, cam in cameras.items():
+            to_write = [cam.id, cam.model, cam.width, cam.height, *cam.params]
+            line = " ".join([str(elem) for elem in to_write])
+            fid.write(line + "\n")
+
+
+def write_cameras_binary(cameras, path_to_model_file):
+    """
+    see: src/base/reconstruction.cc
+        void Reconstruction::WriteCamerasBinary(const std::string& path)
+        void Reconstruction::ReadCamerasBinary(const std::string& path)
+    """
+    with open(path_to_model_file, "wb") as fid:
+        write_next_bytes(fid, len(cameras), "Q")
+        for _, cam in cameras.items():
+            model_id = CAMERA_MODEL_NAMES[cam.model].model_id
+            camera_properties = [cam.id, model_id, cam.width, cam.height]
+            write_next_bytes(fid, camera_properties, "iiQQ")
+            for p in cam.params:
+                write_next_bytes(fid, float(p), "d")
+    return cameras
+
+
+def read_images_text(path):
+    """
+    see: src/base/reconstruction.cc
+        void Reconstruction::ReadImagesText(const std::string& path)
+        void Reconstruction::WriteImagesText(const std::string& path)
+    """
+    images = {}
+    with open(path, "r") as fid:
+        while True:
+            line = fid.readline()
+            if not line:
+                break
+            line = line.strip()
+            if len(line) > 0 and line[0] != "#":
+                elems = line.split()
+                image_id = int(elems[0])
+                qvec = np.array(tuple(map(float, elems[1:5])))
+                tvec = np.array(tuple(map(float, elems[5:8])))
+                camera_id = int(elems[8])
+                image_name = elems[9]
+                elems = fid.readline().split()
+                xys = np.column_stack([tuple(map(float, elems[0::3])), tuple(map(float, elems[1::3]))])
+                point3D_ids = np.array(tuple(map(int, elems[2::3])))
+                images[image_id] = Image(
+                    id=image_id,
+                    qvec=qvec,
+                    tvec=tvec,
+                    camera_id=camera_id,
+                    name=image_name,
+                    xys=xys,
+                    point3D_ids=point3D_ids,
+                )
+    return images
+
+
+def read_images_binary(path_to_model_file):
```

Comment:
i am a bit worried checking this 3rd party in since I don't think it conforms to our python coding bar:
(1) no docstrings
(2) no type hints
(3) no explanation of input or return args.

So in https://github.com/borglab/gtsfm/blob/master/gtsfm/utils/io.py#L234, we have:
```python
def read_images_txt(fpath: str) -> Tuple[Optional[List[Pose3]], Optional[List[str]]]:
    """Read camera poses and image file names from a COLMAP-format images.txt file.
    Reference: https://colmap.github.io/format.html#images-txt
        "The coordinates of the projection/camera center are given by -R^t * T, where
        R^t is the inverse/transpose of the 3x3 rotation matrix composed from the
        quaternion and T is the translation vector. The local camera coordinate system
        of an image is defined in a way that the X axis points to the right, the Y axis
        to the bottom, and the Z axis to the front as seen from the image."
    Args:
        fpath: path to images.txt file
    Returns:
        wTi_list: list of camera poses for each image, or None if file path invalid
        img_fnames: name of image file, for each image, or None if file path invalid
    """
    if not Path(fpath).exists():
        logger.info("%s does not exist", fpath)
        return None, None

    with open(fpath, "r") as f:
        lines = f.readlines()

    wTi_list = []
    img_fnames = []
    # ignore first 4 lines of text -- they contain a description of the file format
    # and a record of the number of reconstructed images.
    for line in lines[4::2]:
        i, qw, qx, qy, qz, tx, ty, tz, i, img_fname = line.split()
        # Colmap provides extrinsics, so must invert
        iRw = Rot3(float(qw), float(qx), float(qy), float(qz))
        wTi = Pose3(iRw, np.array([tx, ty, tz], dtype=np.float64)).inverse()
        wTi_list.append(wTi)
        img_fnames.append(img_fname)

    return wTi_list, img_fnames
```
The only difference there is that we don't return the point3D_ids. But I would prefer we update our internal code to return those, otherwise I think we will have two redundant functions in the codebase

Reply:
Another option would be to put in a PR to COLMAP to bring the 3rd party code up to our standard. That's what we did for PatchmatchNet. But I don't think it resolves the redundancy issue (we would need to maintain two copies of this function)

Reply:
Again, all of the data is in binary. I can add these functions to `io.py` if you'd like. 

However, all of these functions cater to a third-party data format. Unless we're planning on making this our own data format as well, maybe they should be in thirdparty? Or possibly part of the specific loader(s)?

Reply:
@johnwlambert What do you think.

I think having each loader handle reading of their respective data formats makes sense. However, the COLMAP format is pretty widespread, so it could be useful having them in io.py as well.

Reply:
Hi Travis, I couldn't find any binary-encoded data in the COLMAP files. It looks like they parse some of the plain text that way, but it could be done in either way.

I just don't think the COLMAP python code is high-enough quality to check into our repo. I might be ok with the 3rd party route if we can PR into COLMAP and lift the quality of that file to add docstrings, type hints, etc. Otherwise I think its fairly unreadable. But we have different data structures than them, so it might make sense just to add the few missing lines to our files and use our own converter.

Reply:
The AstroNet files use the _COLMAP_ binary format. You can find them in `tests/data/test_2011212_opnav_022`.

Yes, I also don't think the code is high enough quality to check into the repo. I was wondering whether you suggest I add the (properly documented) functions to the individual loader or `io.py`.

Reply:
I see, got it, didn't realize you were using the explicit .bin format. In that case, I would suggest adding a PR to COLMAP to clean it up per our [contributing guidelines](https://github.com/borglab/gtsfm/blob/master/CONTRIBUTING.md), and then we could add it directly to the `thirdparty` dir. That's what we did for PatchmatchNet

Reply:
Redocumented all of the functions. However, I still recommend that we move the modules in `thirdparty/` to [submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules), as they provide an upstream and will likely lead to less licencing issues.